samcli/commands/delete/delete_context.py

```diff
@@ -219,19 +232,23 @@ def delete_ecr_companion_stack(self):
 
             retain_repos = self.ecr_repos_prompts(ecr_companion_stack_template)
 
-            # Delete the repos created by ECR companion stack if not retained
-            ecr_companion_stack_template.delete(retain_resources=retain_repos)
-
             click.echo(f"\t- Deleting ECR Companion Stack {self.companion_stack_name}")
             try:
                 # If delete_stack fails and its status changes to DELETE_FAILED, retain
                 # the user input repositories and delete the stack.
                 self.cf_utils.delete_stack(stack_name=self.companion_stack_name)
                 self.cf_utils.wait_for_delete(stack_name=self.companion_stack_name)
                 LOG.debug("Deleted ECR Companion Stack: %s", self.companion_stack_name)
+
+                # Delete the repos created by ECR companion stack if not retained
+                ecr_companion_stack_template.delete(retain_resources=retain_repos)
             except CfDeleteFailedStatusError:
                 LOG.debug("delete_stack resulted failed and so re-try with retain_resources")
+                # Delete the non retained repos created by ECR companion stack before deleting the stack
+                ecr_companion_stack_template.delete(retain_resources=retain_repos)
+
                 self.cf_utils.delete_stack(stack_name=self.companion_stack_name, retain_resources=retain_repos)
+                self.cf_utils.wait_for_delete(stack_name=self.companion_stack_name)
```

Comment:
sorry I did not get these lines, i have the following questions:
- what is the difference between the call in line 244, and 248 .. for me we are doing the same thing, so what is the value of repeating it.
- why cf_utils.delete_stack does not wait till the stack is really deleted.

Reply:
So the idea was if the stack has termination protection enabled, delete_stack will raise an error and not delete any artifacts. But if it raised CfDeleteFailedStatusError, then delete the artifacts before deleting the stack as there might be repositories which need to be retained.

Reply:
is `termination protection enabled` something that we can know without just trying to delete, and fail.
what should we do in case if SAM failed to delete the ECR Repos

Reply:
describe_stacks api can be used to get this information. I can modify has_stack method in cf_utils.py to check if termination protection is enabled. If it is enabled, raise an exception.